ingest/expression_files/dense_ingestor.py

```diff
@@ -0,0 +1,214 @@
+        IngestFiles.__init__(
+            self, file_path, allowed_file_types=self.ALLOWED_FILE_TYPES
+        )
+        # To allow additional optional keyword arguments like gene_id
+        self.matrix_params = matrix_kwargs
+        self.csv_file_handler, self.file_handler = self.open_file(self.file_path)
+        self.gene_names = {}
+        self.header = DenseIngestor.process_header(next(self.csv_file_handler))
+
+    def execute_ingest(self):
+        # Row after header is needed for R format validation
+        row = next(self.csv_file_handler)
+        if not DenseIngestor.is_valid_format(self.header, row):
+            raise ValueError("Dense matrix has invalid format")
+        # Reset csv reader to first gene row
+        self.csv_file_handler = self.open_file(self.file_path)[0]
+        next(self.csv_file_handler)
+        for gene_docs, data_array_documents in self.transform():
+            self.load(gene_docs, data_array_documents)
+
+    def matches_file_type(file_type):
+        return file_type == "dense"
+
+    @staticmethod
+    def format_gene_name(gene):
+        return gene.strip().strip("'\",")
+
+    @staticmethod
+    def process_header(header):
+        return [value.strip("'\",") for value in header]
+
+    @staticmethod
+    def process_row(row: str):
+        """
+        Performs pre-processing steps for a single row that converts gene
+        scores into floats.
+        """
+
+        def convert_to_float(value: str):
+            # Remove white spaces and quotes
+            value = value.strip("'\",")
+            # Convert string to float and round 3 places
+            return round(float(value), 3)
+
+        result = map(convert_to_float, row[1:])
+        return list(result)
+
+    @staticmethod
+    def filter_expression_scores(scores: List, cells: List):
+        """
+        Filters non-zero gene scores and corresponding cell names
+
+        Returns:
+            tuple (generator):
+                valid_expression_scores (list): non-zero gene scores
+                associated_cells (list): cells
+        """
+        associated_cells = []
+        valid_expression_scores = []
+        for idx, expression_score in enumerate(scores, 1):
+            if (
+                expression_score != 0
+                and expression_score is not None
+                and not str(expression_score).isspace()
+                and expression_score != ""
+            ):
+                valid_expression_scores.append(expression_score)
+                associated_cells.append(cells[idx])
+        return valid_expression_scores, associated_cells
+
+    @staticmethod
+    def is_valid_format(header, row):
+        return all(
+            [
+                DenseIngestor.has_unique_header(header),
+                DenseIngestor.has_gene_keyword(header, row),
+                DenseIngestor.header_has_valid_values(header),
+            ]
+        )
+
+    @staticmethod
+    def has_unique_header(header: List):
```

Comment:
This [ticket](SCP-2635 Ensure cell names from expression matrices are unique across files) tracks the issues

Reply:
Hmm... yeah, that's tricky.  I would lean towards code just querying for the cell names based on the study_id that it knows.  It might require going down a couple of layers in query, but our python code has to know about those mongo structures since it's doing all the ingest that put those cell names there in the first place.

Reply:
Thanks for creating a ticket to make sure that unique cells will be checked. As long as we have planned follow-on work, that satisfies my change request.